### PR TITLE
fix(TCOMP-236): fix input component archetype

### PR DESCRIPTION
**Pull request by gonchar-ivan**
**Please check if the PR fulfills these requirements**

- [x] The commit message follows Talend standard
- [x] Tests for the changes have been added (for bug fixes / features)

**What is the current behavior?** (You can also link to an open issue here)

input-components-archetype is not correct

**What is the new behavior?**

archetype was cleaned up and fixed

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:
https://jira.talendforge.org/browse/TCOMP-236

It's better to check generated project rather that this PR:
https://github.com/Talend/components/compare/work/TCOMP-236_input_archetype_example?expand=1

or in branch
https://github.com/Talend/components/tree/work/TCOMP-236_input_archetype_example/components/components-table

If you want try to build archetype do following:
mvn clean install (on components\examples\components-api-archetypes\input-component-archetypeinput-component)

than use following command to generate project:
mvn org.apache.maven.plugins:maven-archetype-plugin:2.4:generate -DarchetypeGroupId=org.talend.components -DarchetypeArtifactId=input-component-archetypes -DarchetypeVersion=0.18.0-SNAPSHOT

### CODE REVIEW

**Comment by build-talend-common** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/192/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/192/)

**Comment by build-talend-common** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/283/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/283/)

**Comment by gonchar-ivan** [comment]
Build failed not because of my changes

**Comment by build-talend-common** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/components_pull-request_master/357/) [:star:](https://ci-common.datapwn.com/job/components_pull-request_master/357/)

File: examples/components-api-archetypes/input-component-archetype/src/main/resources/archetype-resources/__rootArtifactId__-runtime/src/main/java/runtime/reader/__componentName__Reader.java [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+
+    @Override
+    public boolean advance() throws IOException {
+        if (!started) {
+            throw new IllegalStateException("Reader wasn't started");
+        }
+        hasMore = reader.ready();
+        if (hasMore) {
+            String line = reader.readLine();
+            IndexedRecord dataRecord = getConverter(line).convertToAvro(line);
```
Comment (sgandon) [inline examples/components-api-archetypes/input-component-archetype/src/main/resources/archetype-resources/__rootArtifactId__-runtime/src/main/java/runtime/reader/__componentName__Reader.java]: this is strange here.
I would have imagined you'd do something like (some pseudo code)
```
//create the data schema
Schema dataSchema = getRecordSchema(line); <- create the schema only once for all the records (some code removed here)
//create the data IndexRecord
IndexRecord dataRecord = getConverter(recordSchema).convertToAvro(line);
//create the outOfboand record (since the schema is static)
IndexedRecord outOfBandRecord = new GenericData.Record(${componentName}Properties.outOfBandSchema);
outOfBandRecord.put(0, result.totalCount);
then create the root IndexRecord
current = OutOfBandUtils.createRootRecord(dataRecord, outOfBandRecord);
```
To looks me this is more clear to me. And the `OutOfBandUtils.createRootRecord` would be in daikon too.
What do you think ?

File: examples/components-api-archetypes/input-component-archetype/src/main/resources/archetype-resources/__rootArtifactId__-runtime/src/test/resources/log4j.properties [hunk 2]
```diff
@@ -0,0 +1,7 @@
+log4j.rootLogger=DEBUG, stdout
```
Comment (sgandon) [inline examples/components-api-archetypes/input-component-archetype/src/main/resources/archetype-resources/__rootArtifactId__-runtime/src/test/resources/log4j.properties]: could you set it to INFO instead of DEBUG this is causing too much un-useful  logs in the build.

File: examples/components-api-archetypes/input-component-archetype/src/main/resources/archetype-resources/__rootArtifactId__-definition/src/main/java/__componentPackage__/__componentName__Properties.java [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+     * 
+     * Main schema could be changes by user. So method should reconstruct Root schema each time
+     * 
+     * @param connector token to get the associated schema
+     * @param isOutputConnection whether the connection is an outgoing or incoming one
+     * @return Root schema, which contains main schema related to connector and static Out of band schema 
+     */
+    @Override
+    public Schema getSchema(Connector connector, boolean isOutputConnection) {
+        // design-time main schema associated with specified connector
```
Comment (gonchar-ivan) [inline examples/components-api-archetypes/input-component-archetype/src/main/resources/archetype-resources/__rootArtifactId__-definition/src/main/java/__componentPackage__/__componentName__Properties.java]: hm... I just realized _isOutputConnection_ parameter is not used here (and it is not required).
Schema fully defined by connector parameter (Input and Output schema have different connector instances associated with)
_isOutputConnection_ is used in getAvailableConnectors() to specify which kind of connector caller requested.
However, I'll add if statement for the moment

File: examples/components-api-archetypes/input-component-archetype/src/main/resources/archetype-resources/__rootArtifactId__-runtime/pom.xml [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+		<!-- APIs for all Components -->
+		<dependency>
+			<groupId>org.talend.components</groupId>
+			<artifactId>components-api</artifactId>
+			<version>${project.parent.version}</version>
+		</dependency>
+
+		<!-- Shared component properties such as Use/Password or Schema -->
+		<dependency>
```
Comment (gonchar-ivan) [inline examples/components-api-archetypes/input-component-archetype/src/main/resources/archetype-resources/__rootArtifactId__-runtime/pom.xml]: Agree about components-common. It is not really used in FileInput Runtime module.

And I'm not sure about components-api. My point of view is if some module directly uses some dependency (runtime module imports api classes), this dependency (even transitive) should be explicitly declared in module's pom.xml 
And runtime module directly uses components-api dependency.

Found following answers about the topic:
http://stackoverflow.com/questions/4226756/maven-should-i-keep-or-remove-declared-dependencies-that-are-also-transitives
http://stackoverflow.com/questions/4226756/maven-should-i-keep-or-remove-declared-dependencies-that-are-also-transitives

File: examples/components-api-archetypes/input-component-archetype/src/main/resources/archetype-resources/__rootArtifactId__-runtime/src/main/java/avro/StringConverter.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
+// ============================================================================
```
Comment (gonchar-ivan) [inline examples/components-api-archetypes/input-component-archetype/src/main/resources/archetype-resources/__rootArtifactId__-runtime/src/main/java/avro/StringConverter.java]: 1) My first idea was that each component has a set of converters which are specific only for that component.
However, seems I was wrong. My statement is reasonable for top-level converter (top-level means datum object to IndexedRecord), but converters for each field may have similar logic. So, I agree to move them to daikon 

2) Just a note about package naming conventions in Daikon.
For converters we may have 2 approaches:
org/talend/daikon/avro/converter/{datumType} e.g org/talend/daikon/avro/converter/string
org/talend/daikon/avro/converters/{avroType} e.g. org/talend/daikon/avro/converter/logicalTime

first package name means this package it contains converters from datum string to (all) avro types
second package name means this package it contains converters from any datum type to avro logicalTime

I don't know which approach is best:
in datum approach we may have potentially infinite number of datum types -> thus packages
(however, surely we will not put all converters to daikon, because I think some of them will be really component specific)

File: examples/components-api-archetypes/input-component-archetype/src/main/resources/archetype-resources/__rootArtifactId__-runtime/src/test/resources/log4j.properties (hunk 2 above)
Comment (gonchar-ivan) [inline examples/components-api-archetypes/input-component-archetype/src/main/resources/archetype-resources/__rootArtifactId__-runtime/src/test/resources/log4j.properties]: done in both log4j.properties: definition and runtime (will be done for integration module too, if such module required)

File: examples/components-api-archetypes/input-component-archetype/src/main/resources/archetype-resources/__rootArtifactId__-runtime/src/main/java/runtime/reader/__componentName__Reader.java (hunk 1 above)
Comment (gonchar-ivan) [inline examples/components-api-archetypes/input-component-archetype/src/main/resources/archetype-resources/__rootArtifactId__-runtime/src/main/java/runtime/reader/__componentName__Reader.java]: Several notes:
1) Agree, Root record creation is a little bit dirty. So, better to have 1 code line instead of following 3:
```
current = new GenericData.Record(getRootSchema(line));
current.put(0, dataRecord);
current.put(1, outOfBandRecord);
```
2) However, I don't want to recreate Root Schema each time. I want to be able to cache it.
OutOfBandUtils.createRootRecord(dataRecord, outOfBandRecord) call supposes Root Schema will be recreated for each new record (from dataRecord, outOfBandRecord schemas)
3) As I remember, we discussed that in future we may have something else besides data record and outOfBand record inside root record (3 and more IndexedRecords inside one Root record). So, as for me this Util class should have RootRecordUtils name.
4) Another idea is to extend IndexedRecord. We may implement _RootRecord extends IndexedRecord_.
Its constructor should accept Root Schema and all nested records (data and outOfBand for the moment)
5) Why daikon and not components-api? I thought IndexedRecords are used only in components.
(I realized I made a mistake when I put RootSchema/RootRecordUtils in component-common. So, I think I should move to components-api, but I'm not sure it is needed in daikon)

**Comment by build-talend-common** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/75/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/75/)

**Comment by build-talend-common** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/16/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/16/)

**Comment by build-talend-common** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/17/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/17/)

File: examples/components-api-archetypes/input-component-archetype/src/main/resources/archetype-resources/__rootArtifactId__-definition/src/main/java/StringDelimiter.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+package ${package};
+
+/**
+ * Enumeration of standard string delimiters
+ * 
+ * This shows possible values for dropdown list UI element i18n message for it
+ * should be placed in TableInputProperties.properties file
+ */
+public enum StringDelimiter {
+	SEMICOLON {
```
Comment (sgandon) [inline examples/components-api-archetypes/input-component-archetype/src/main/resources/archetype-resources/__rootArtifactId__-definition/src/main/java/StringDelimiter.java]: You could make it more simple like
SEMINCOLON(";"), COLON(",")...
private delim;
public StringDelimiter(String delim){
    this.delim = delim;
}
public getDelimiter(){
    return delim;
}

Comment (gonchar-ivan) [inline examples/components-api-archetypes/input-component-archetype/src/main/resources/archetype-resources/__rootArtifactId__-definition/src/main/java/StringDelimiter.java]: got it. Thanks. fixing

**Comment by build-talend-common** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/19/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/19/)

File: examples/components-api-archetypes/input-component-archetype/src/test/resources/projects/archetypeTest/goal.txt [hunk 7]
```diff
@@ -1 +1 @@
-clean verify clean
+clean install clean
```
Comment (sgandon) [inline examples/components-api-archetypes/input-component-archetype/src/test/resources/projects/archetypeTest/goal.txt]: why did you change from verify to install ? this is the maven goal launched when testing the archetype creation ?

File: examples/components-api-archetypes/input-component-archetype/src/main/resources/archetype-resources/__rootArtifactId__-definition/src/main/java/RuntimeInfoProvider.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+package ${package};
+
+import org.talend.components.api.component.runtime.DependenciesReader;
+import org.talend.components.api.component.runtime.JarRuntimeInfo;
+import org.talend.daikon.runtime.RuntimeInfo;
+
+/**
+ * Service class, which provides {@link RuntimeInfo} instances
+ */
+public final class RuntimeInfoProvider {
```
Comment (sgandon) [inline examples/components-api-archetypes/input-component-archetype/src/main/resources/archetype-resources/__rootArtifactId__-definition/src/main/java/RuntimeInfoProvider.java]: I was wondering if we really need a class of this, I would rather be using the FamillyDefinition class.

File: examples/components-api-archetypes/input-component-archetype/src/test/resources/projects/archetypeTest/goal.txt (hunk 7 above)
Comment (gonchar-ivan) [inline examples/components-api-archetypes/input-component-archetype/src/test/resources/projects/archetypeTest/goal.txt]: this goal is executed on project generated from archetype.
verify -> install for integration tests. Otherwise IT in OSGi container fails

File: examples/components-api-archetypes/input-component-archetype/src/main/resources/archetype-resources/__rootArtifactId__-definition/src/main/java/RuntimeInfoProvider.java (hunk 8 above)
Comment (gonchar-ivan) [inline examples/components-api-archetypes/input-component-archetype/src/main/resources/archetype-resources/__rootArtifactId__-definition/src/main/java/RuntimeInfoProvider.java]: Definition classes uses Properties classes
see getPropertyClass() method

When I add some design time feature and need to get RuntimeInfo in Properties class.
In case of using FamilyDefinition as RuntimeInfo provider, Properties class will use FamilyDefinition.

Thus, something like cycle dependency (however, inside package or module only).

I wanted to avoid this.
I saw both Definition and Properties classes require RuntimeInfo.
Thus, I introduce separate class, which provides it.
I'm not sure about this decision.

**Comment by build-talend-common** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/components_pull-request_master/85/) [:star:](https://ci-common.datapwn.com/job/components_pull-request_master/85/)
